File: src/s2/convert.ts
import { fromST, toLonLat } from './s2Point';

import type { Face, S2Feature, VectorFeature, VectorGeometry, VectorPoint } from '../';

/**
 * Convert an S2Feature to a GeoJSON Feature
 * @param data - S2Feature
 * @returns - GeoJSON Feature
 */
export function toWM(data: S2Feature): VectorFeature {
  const { id, face, properties, metadata, geometry } = data;
  convertGeometry(face, geometry);
  return {
    id,
    type: 'VectorFeature',
    properties,
    metadata,
    geometry,
  };
}

/**
 * Underlying conversion mechanic to move S2Geometry to GeoJSON Geometry
 * @param face - Face
 * @param geometry - S2 Geometry
 */
function convertGeometry(face: Face, geometry: VectorGeometry): void {
  const { type, coordinates } = geometry;
  if (type === 'Point') convertGeometryPoint(face, coordinates);
  else if (type === 'MultiPoint') coordinates.forEach((point) => convertGeometryPoint(face, point));
  else if (type === 'LineString') coordinates.forEach((point) => convertGeometryPoint(face, point));
  else if (type === 'MultiLineString')
    coordinates.forEach((line) => line.forEach((point) => convertGeometryPoint(face, point)));
  else if (type === 'Polygon')
    coordinates.forEach((line) => line.forEach((point) => convertGeometryPoint(face, point)));
  else if (type === 'MultiPolygon')
    coordinates.forEach((polygon) =>
      polygon.forEach((line) => line.forEach((point) => convertGeometryPoint(face, point))),
    );
  else {
    throw new Error('Invalid S2Geometry type');
  }
}

/**
 * Mutate an S2 Point to a GeoJSON Point
 * @param face - Face
 * @param point - S2 Point
 */
function convertGeometryPoint(face: Face, point: VectorPoint): void {
  const { x: s, y: t } = point;
  const [lon, lat] = toLonLat(fromST(face, s, t));
  point.x = lon;
  point.y = lat;
}

File: rust/s2/convert.rs
use crate::{Face, LonLat, S2CellId, VectorFeature, VectorGeometry, VectorPoint};

impl<M: Clone> VectorFeature<M> {
    /// Convert an S2 Feature to a GeoJSON Vector Feature
    pub fn to_wm(&self) -> Self {
        if self._type == "VectorFeature" {
            return self.clone();
        }
        let mut geometry = self.geometry.clone();
        convert_geometry(self.face, &mut geometry);
        VectorFeature::<M>::new_wm(
            self.id,
            self.properties.clone(),
            geometry,
            self.metadata.clone(),
        )
    }
}

/// Underlying conversion mechanic to move S2Geometry to GeoJSON Geometry
fn convert_geometry(face: Face, geometry: &mut VectorGeometry) {
    match geometry {
        VectorGeometry::Point(point) => convert_geometry_point(face, &mut point.coordinates),
        VectorGeometry::LineString(points) | VectorGeometry::MultiPoint(points) => {
            points.coordinates.iter_mut().for_each(|point| convert_geometry_point(face, point))
        }
        VectorGeometry::Polygon(lines) | VectorGeometry::MultiLineString(lines) => lines
            .coordinates
            .iter_mut()
            .for_each(|line| line.iter_mut().for_each(|point| convert_geometry_point(face, point))),
        VectorGeometry::MultiPolygon(polygons) => {
            polygons.coordinates.iter_mut().for_each(|polygon| {
                polygon.iter_mut().for_each(|line| {
                    line.iter_mut().for_each(|point| convert_geometry_point(face, point))
                })
            })
        }
    }
}

/// Mutate an S2 Point to a GeoJSON Point
fn convert_geometry_point(face: Face, point: &mut VectorPoint) {
    let LonLat { lon, lat } = (&S2CellId::from_face_st(face.into(), point.x, point.y)).into();
    point.x = lon;
    point.y = lat;
}
